File: apps/web/src/lib/api/employer.ts
/**
 * Typed API functions for employer-side endpoints.
 * Mirrors the Pydantic schemas in apps/api/app/schemas/employer.py.
 *
 * Safety: ApplicantMatchSummary exposes only safe fields —
 * no user_id, no email, no admin-only data.
 */
import { apiFetch } from "./client";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface EmployerCompanySummary {
  employer_id: string;
  name: string;
  industry: string | null;
  city: string | null;
  state: string | null;
  is_partner: boolean;
  total_jobs: number;
  active_jobs: number;
  accepts_internal_applications_default: boolean;
}

export interface EmployerJobSummary {
  job_id: string;
  title: string;
  city: string | null;
  state: string | null;
  work_setting: string | null;
  is_active: boolean;
  posted_date: string | null;
  created_at: string;
  total_visible: number;
  eligible_count: number;
  near_fit_count: number;
  family_code: string | null;
  family_name: string | null;
  employment_type: string | null;
  pay_min: number | null;
  pay_max: number | null;
  pay_type: string | null;
  pay_raw: string | null;
  source: string | null;
  source_site: string | null;
  is_stale: boolean;
  apply_link_status: string | null;
  /** Lifecycle: active | paused | filled | closed (is_active ⇔ 'active'). */
  status: JobLifecycleStatus;
  previous_status: JobLifecycleStatus | null;
  /** True when the job has applications/interest/outreach/hire history —
   *  Delete is refused (Close is the honest action). */
  has_activity: boolean;
  /** Career-source freshness (source-synced jobs only). */
  source_last_seen_at: string | null;
  source_vanished_at: string | null;
}

export type JobLifecycleStatus = "active" | "paused" | "filled" | "closed";

export interface JobStatusOut {
  job_id: string;
  status: JobLifecycleStatus;
  previous_status: JobLifecycleStatus | null;
  is_active: boolean;
}

export interface EmployerJobFacets {
  families: { value: string; label: string }[];
  states: string[];
  sources: string[];
  employment_types: string[];
}

export interface EmployerJobsListResponse {
  employer_id: string;
  company_name: string;
  jobs: EmployerJobSummary[];
  total_jobs: number;
  unfiltered_total: number | null;
  supports_internal_apply: boolean;
  facets: EmployerJobFacets;
}

export interface EmployerJobFilters {
  q?: string;
  families?: string;
  states?: string;
  city?: string;
  employment_types?: string;
  sources?: string;
  status?: string;          // active | inactive | stale
  apply_link?: string;      // ok | broken | unchecked
  has_pay?: boolean;
  pay_gte?: number;
  internal_apply?: boolean;
  posted_from?: string;
  posted_to?: string;
  candidates?: string;      // none | 1_9 | 10_49 | over_50
  sort?: string;            // newest | posted | title | pay
}

export interface ApplicantMatchSummary {
  match_id: string;
  applicant_id: string;
  first_name: string | null;
  last_name: string | null;
  city: string | null;
  state: string | null;
  region: string | null;
  willing_to_relocate: boolean;
  willing_to_travel: boolean;
  program_name_raw: string | null;
  canonical_job_family_code: string | null;
  expected_completion_date: string | null;
  available_from_date: string | null;
  eligibility_status: "eligible" | "near_fit";
  match_label: "strong_fit" | "good_fit" | "moderate_fit" | "low_fit" | null;
  policy_adjusted_score: number | null;
  top_strengths: string[];
  top_gaps: string[];
  recommended_next_step: string | null;
  confidence_level: "high" | "medium" | "low" | null;
  requires_review: boolean;
  geography_note: string | null;
  applicant_interest: "interested" | "applied" | "not_interested" | null;
}

export interface RankedApplicantsResponse {
  job_id: string;
  job_title: string;
  employer_name: string;
  total_visible: number;
  eligible_count: number;
  near_fit_count: number;
  applicants: ApplicantMatchSummary[];
  filter_eligibility: string | null;
  filter_min_score: number | null;
  filter_state: string | null;
  filter_willing_to_relocate: boolean | null;
  /** Pagination over the FILTERED list. */
  filtered_total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface ApplicantFilters {
  eligibility?: "all" | "eligible" | "near_fit";
  minScore?: number;
  state?: string;
  willingToRelocate?: boolean;
  /** Partial, case-insensitive name search — narrows live as the user types. */
  q?: string;
  /** 1-based page over the filtered list. */
  page?: number;
  perPage?: number;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function fetchMyCompany(
  token: string,
): Promise<EmployerCompanySummary> {
  return apiFetch<EmployerCompanySummary>("/employer/me/company", token);
}

/** Employer-editable company settings (partial update). */
export async function patchMyCompany(
  token: string,
  payload: { accepts_internal_applications_default: boolean },
): Promise<EmployerCompanySummary> {
  return apiFetch<EmployerCompanySummary>("/employer/me/company", token, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export async function fetchMyJobs(
  token: string,
  filters: EmployerJobFilters = {},
): Promise<EmployerJobsListResponse> {
  const qs = new URLSearchParams();
  for (const [k, v] of Object.entries(filters)) {
    if (v !== undefined && v !== "") qs.set(k, String(v));
  }
  const q = qs.toString();
  return apiFetch<EmployerJobsListResponse>(`/employer/me/jobs${q ? `?${q}` : ""}`, token);
}

/** Move a job through its lifecycle. 409 = illegal transition or a
 *  concurrent change; the caller should refresh. */
export async function patchJobStatus(
  token: string,
  jobId: string,
  status: JobLifecycleStatus,
): Promise<JobStatusOut> {
  return apiFetch<JobStatusOut>(`/employer/me/jobs/${jobId}/status`, token, {
    method: "PATCH",
    body: JSON.stringify({ status }),
  });
}

/** REAL undo for the last lifecycle transition (restores previous_status). */
export async function revertJobStatus(
  token: string,
  jobId: string,
): Promise<JobStatusOut> {
  return apiFetch<JobStatusOut>(`/employer/me/jobs/${jobId}/status/revert`, token, {
    method: "POST",
  });
}

/** Delete a zero-activity posting. 409 when it has history — close instead. */
export async function deleteJob(token: string, jobId: string): Promise<void> {
  await apiFetch<{ ok: boolean }>(`/employer/me/jobs/${jobId}`, token, {
    method: "DELETE",
  });
}

export async function fetchJobApplicants(
  jobId: string,
  token: string,
  filters?: ApplicantFilters,
): Promise<RankedApplicantsResponse> {
  const params = new URLSearchParams();
  if (filters?.eligibility && filters.eligibility !== "all") {
    params.set("eligibility", filters.eligibility);
  }
  if (filters?.minScore && filters.minScore > 0) {
    params.set("min_score", String(filters.minScore));
  }
  if (filters?.state) {
    params.set("state", filters.state);
  }
  if (filters?.willingToRelocate !== undefined) {
    params.set("willing_to_relocate", String(filters.willingToRelocate));
  }
  if (filters?.q) {
    params.set("q", filters.q);
  }
  if (filters?.page && filters.page > 1) {
    params.set("page", String(filters.page));
  }
  if (filters?.perPage) {
    params.set("per_page", String(filters.perPage));
  }
  const qs = params.toString();
  const path = `/employer/me/jobs/${jobId}/applicants${qs ? `?${qs}` : ""}`;
  return apiFetch<RankedApplicantsResponse>(path, token);
}

// ---------------------------------------------------------------------------
// Display helpers
// ---------------------------------------------------------------------------

export function formatWorkSetting(ws: string | null): string {
  switch (ws) {
    case "remote":   return "Remote";
    case "hybrid":   return "Hybrid";
    case "on_site":  return "On-site";
    case "flexible": return "Flexible";
    default:         return ws ?? "—";
  }
}

export function formatAvailability(
  availableFrom: string | null,
  expectedCompletion: string | null,
): string {
  const date = availableFrom ?? expectedCompletion;
  if (!date) return "Not set";
  try {
    return new Date(date).toLocaleDateString("en-US", {
      month: "short",
      year: "numeric",
    });
  } catch {
    return date;
  }
}

export function formatApplicantName(
  first: string | null,
  last: string | null,
): string {
  return [first, last].filter(Boolean).join(" ") || "Anonymous";
}

File: apps/api/app/schemas/employer.py
"""
Pydantic schemas for employer-facing API endpoints.
Mirrors the employer-side DB shape.

Safety rules (DECISIONS.md):
  - ApplicantMatchSummary exposes only safe summary fields (no user_id, no email,
    no admin-only notes, no other employers' data)
  - employer_global_candidate_search_default: false
"""
from __future__ import annotations

from pydantic import BaseModel, Field, field_validator, model_validator

# Enum-ish string fields validated at the schema edge so bad values 422
# instead of exploding inside SQL casts.
_WORK_SETTINGS = {"remote", "hybrid", "on_site", "flexible"}
_TRAVEL_REQS = {"none", "light", "moderate", "frequent"}
_PAY_TYPES = {"hourly", "annual", "contract"}
_EXPERIENCE_LEVELS = {"entry", "mid", "senior"}
# Profile-sourced groups an employer may require at internal apply time.
_PROFILE_FIELD_GROUPS = {"contact", "location", "program", "availability", "credentials", "resume"}


def _check_choice(value: str | None, allowed: set[str], field_name: str) -> str | None:
    if value is None:
        return None
    v = value.strip().lower()
    if v not in allowed:
        raise ValueError(f"{field_name} must be one of: {', '.join(sorted(allowed))}")
    return v

# ---------------------------------------------------------------------------
# Company / employer
# ---------------------------------------------------------------------------

class EmployerCompanySummary(BaseModel):
    employer_id: str
    name: str
    industry: str | None
    city: str | None
    state: str | None
    is_partner: bool
    total_jobs: int
    active_jobs: int
    # Company-wide default for jobs that don't set their own flag.
    accepts_internal_applications_default: bool = False


class CompanySettingsPatch(BaseModel):
    """Employer-editable company settings (partial update)."""
    accepts_internal_applications_default: bool | None = None


# ---------------------------------------------------------------------------
# Jobs
# ---------------------------------------------------------------------------

class EmployerJobSummary(BaseModel):
    job_id: str
    title: str
    city: str | None
    state: str | None
    work_setting: str | None
    is_active: bool
    posted_date: str | None
    created_at: str
    total_visible: int     # visible-to-employer matches (pre-filter)
    eligible_count: int
    near_fit_count: int
    # Granular-filter columns (additive; defaults keep old consumers working)
    family_code: str | None = None
    family_name: str | None = None
    employment_type: str | None = None
    pay_min: float | None = None
    pay_max: float | None = None
    pay_type: str | None = None
    pay_raw: str | None = None
    source: str | None = None
    source_site: str | None = None
    is_stale: bool = False
    apply_link_status: str | None = None
    # Lifecycle (additive): active | paused | filled | closed. is_active stays
    # the machine truth for visibility (status='active' ⇔ is_active=TRUE).
    status: str = "active"
    previous_status: str | None = None
    # Activity gate for Delete (else Close is the honest action).
    has_activity: bool = False
    # Career-source freshness (source-synced jobs only): when the posting was
    # last seen on the employer's own careers site, and when it vanished.
    source_last_seen_at: str | None = None
    source_vanished_at: str | None = None


class EmployerJobFacets(BaseModel):
    """Option lists drawn from THIS employer's own jobs only."""
    families: list[dict[str, str]] = []      # {value, label}
    states: list[str] = []
    sources: list[str] = []
    employment_types: list[str] = []


class EmployerJobsListResponse(BaseModel):
    employer_id: str
    company_name: str
    jobs: list[EmployerJobSummary]
    total_jobs: int
    # Filtered-list metadata (additive)
    unfiltered_total: int | None = None
    supports_internal_apply: bool = False
    facets: EmployerJobFacets = EmployerJobFacets()


class _JobFieldRules(BaseModel):
    """Shared cross-field/enum rules for job create + update payloads."""

    @field_validator("title_raw", check_fields=False)
    @classmethod
    def _title_not_blank(cls, v: str | None) -> str | None:
        if v is None:
            return None
        v = v.strip()
        if not v:
            raise ValueError("title_raw must not be blank")
        return v

    @field_validator("work_setting", check_fields=False)
    @classmethod
    def _work_setting_valid(cls, v: str | None) -> str | None:
        return _check_choice(v, _WORK_SETTINGS, "work_setting")

    @field_validator("travel_requirement", check_fields=False)
    @classmethod
    def _travel_valid(cls, v: str | None) -> str | None:
        return _check_choice(v, _TRAVEL_REQS, "travel_requirement")

    @field_validator("pay_type", check_fields=False)
    @classmethod
    def _pay_type_valid(cls, v: str | None) -> str | None:
        return _check_choice(v, _PAY_TYPES, "pay_type")

    @field_validator("experience_level", check_fields=False)
    @classmethod
    def _experience_valid(cls, v: str | None) -> str | None:
        return _check_choice(v, _EXPERIENCE_LEVELS, "experience_level")

    @field_validator("required_profile_fields", check_fields=False)
    @classmethod
    def _profile_fields_valid(cls, v: list[str] | None) -> list[str] | None:
        if v is None:
            return None
        cleaned: list[str] = []
        for item in v:
            key = item.strip().lower()
            if key not in _PROFILE_FIELD_GROUPS:
                raise ValueError(
                    f"required_profile_fields entries must be one of: {', '.join(sorted(_PROFILE_FIELD_GROUPS))}"
                )
            if key not in cleaned:
                cleaned.append(key)
        return cleaned

    @model_validator(mode="after")
    def _pay_range_ordered(self):
        pay_min = getattr(self, "pay_min", None)
        pay_max = getattr(self, "pay_max", None)
        if pay_min is not None and pay_max is not None and pay_min > pay_max:
            raise ValueError("pay_min cannot be greater than pay_max")
        return self


class JobCreateRequest(_JobFieldRules):
    title_raw: str = Field(min_length=1, max_length=200)
    city: str | None = Field(default=None, max_length=120)
    state: str | None = Field(default=None, max_length=50)
    work_setting: str | None = None   # remote | hybrid | on_site | flexible
    travel_requirement: str | None = None  # none | light | moderate | frequent
    pay_min: float | None = Field(default=None, ge=0)
    pay_max: float | None = Field(default=None, ge=0)
    pay_type: str | None = None       # hourly | annual | contract
    description_raw: str | None = Field(default=None, max_length=20000)
    requirements_raw: str | None = Field(default=None, max_length=20000)
    experience_level: str | None = None   # entry | mid | senior
    # Internal-apply configuration
    accepts_internal_applications: bool | None = None
    required_profile_fields: list[str] | None = Field(default=None, max_length=6)


class JobUpdateRequest(_JobFieldRules):
    """All fields optional — only provided fields are updated (PATCH semantics)."""
    title_raw: str | None = Field(default=None, min_length=1, max_length=200)
    city: str | None = Field(default=None, max_length=120)
    state: str | None = Field(default=None, max_length=50)
    work_setting: str | None = None
    travel_requirement: str | None = None
    pay_min: float | None = Field(default=None, ge=0)
    pay_max: float | None = Field(default=None, ge=0)
    pay_type: str | None = None
    description_raw: str | None = Field(default=None, max_length=20000)
    requirements_raw: str | None = Field(default=None, max_length=20000)
    experience_level: str | None = None
    is_active: bool | None = None
    accepts_internal_applications: bool | None = None
    required_profile_fields: list[str] | None = Field(default=None, max_length=6)


class JobCreateResponse(BaseModel):
    job_id: str
    title_raw: str
    is_active: bool
    created_at: str


class JobDetail(BaseModel):
    """Full job detail — used to pre-fill the edit form."""
    job_id: str
    title_raw: str
    city: str | None = None
    state: str | None = None
    work_setting: str | None = None
    travel_requirement: str | None = None
    pay_min: float | None = None
    pay_max: float | None = None
    pay_type: str | None = None
    description_raw: str | None = None
    requirements_raw: str | None = None
    experience_level: str | None = None
    is_active: bool
    # Internal-apply configuration. `accepts_internal_applications` is the raw
    # per-job flag (None = inherit); `internal_apply_effective` folds in the
    # company default so the UI shows the truth.
    accepts_internal_applications: bool | None = None
    required_profile_fields: list[str] = Field(default_factory=list)
    internal_apply_effective: bool = False


# ---------------------------------------------------------------------------
# Applicant match summary (employer view)
# Safe fields only — no user_id, no email, no admin-only fields.
# ---------------------------------------------------------------------------

class ApplicantMatchSummary(BaseModel):
    match_id: str
    applicant_id: str          # internal applicant UUID (not auth user_id)
    first_name: str | None
    last_name: str | None
    city: str | None
    state: str | None
    region: str | None
    willing_to_relocate: bool
    willing_to_travel: bool
    program_name_raw: str | None
    canonical_job_family_code: str | None
    expected_completion_date: str | None
    available_from_date: str | None
    eligibility_status: str              # eligible | near_fit
    match_label: str | None              # strong_fit | good_fit | moderate_fit | low_fit
    policy_adjusted_score: float | None
    top_strengths: list[str]
    top_gaps: list[str]
    recommended_next_step: str | None
    confidence_level: str | None         # high | medium | low
    requires_review: bool
    geography_note: str | None
    applicant_interest: str | None       # interested | applied | not_interested | None


class RankedApplicantsResponse(BaseModel):
    job_id: str
    job_title: str
    employer_name: str
    # Total counts (pre-filter, for dashboard display)
    total_visible: int
    eligible_count: int
    near_fit_count: int
    # Filtered, ranked applicant list
    applicants: list[ApplicantMatchSummary]
    # Applied filter context (for UI to reflect active filters)
    filter_eligibility: str | None
    filter_min_score: float | None
    filter_state: str | None
    filter_willing_to_relocate: bool | None
    # Pagination over the FILTERED list (defaults keep old clients working)
    filtered_total: int = 0
    page: int = 1
    per_page: int = 25
    total_pages: int = 1
